design/re-introspection-interface.md

```diff
@@ -0,0 +1,76 @@
+# re-introspection API
+
+## Status
+
+implementable
+
+## Summary
+
+A declarative API is proposed to request the baremetal operator to
+re-introspect a `Ready` BareMetalHost.
+
+## Motivation
+
+We would like to have an interface to allow a user to re-gather hardware
+inventory of a `Ready` BareMetalHost when a hardware replacement is made.
+When a user of the underlying infrastructure makes some changes to the actual
+server (e.g. replace or add NIC, disk, etc.), the latest hardware inventory
+including those changes need to be re-collected and updated on the spec of the
+corresponding BareMetalHost object without having to delete it.
+
+Implementation of this proposal is based on using annotation (similar to
+[Reboot API](https://github.com/metal3-io/metal3-docs/blob/master/design/reboot-interface.md))
+to request re-introspection of a `Ready` BareMetalHost.
+Once the annotation is set on BareMetalHost, the baremetal operator will
+request hardware introspection of the host from
+Ironic.
+
+### Goals
+
+- A declarative API to perform re-introspection
+- Use this API for future Metal³ remediation controller
+
+### Non-Goals
+
+- Automated mechanism to trigger re-introspection
+
+## Proposal
+
+We follow Reboot API implementation and expect to implement the similar
+annotation based interface for a user to trigger re-introspection.
+
+`reintrospection.metal3.io` annotation form on BareMetalHost object
+triggers the controller to query Ironic for introspection of a host. This form
+has set-and-forget semantics and controller removes the annotation once
+introspection is completed.
+While the host is being re-introspected, the BareMetalHost will stay in
+`Inspecting` state until the process is completed.
+
+Re-introspection API can be requested only when BareMetalHost is in `Ready`
+state. If a re-introspection request is made while BareMetalHost is any other
+state than `Ready`, the request will be ignored, i.e., annotation will
+not be set on BareMetalHost object. This is important in order to not reboot a
+BareMetalHost (e.g. when `spec.provisioning.state == provisioned`) or avoid
+having unintended inconsistent states.
+
+After completing re-introspection, previous inspection data should be updated
+both in Ironic and on the spec of the BareMetalHost object. Both
+`status.operationHistory.inspect.start` and
+`status.operationHistory.inspect.end` timestamps should be updated accordingly.
+
+### User stories
+
+#### Story1
+
+- As a cluster admin, I would like to have a simple way of triggering
+  introspection for my server after I replace NIC, disk, etc.
```

Comment:
Another approach we have discussed for keeping the hardware details up to date is to run introspection from a pod running on the node. You could include that in the alternatives section of this document.

Reply:
That is an interesting approach. However it would work only if the hardware change is not breaking. In case the user would change the ethernet card, the mac address would change. Considering that the network configuration of the host is based on the old mac address, the node might not come back up, and the daemon set would then not run and be unable to report the new hardware. We probably still need to introspect with IPA in some corner cases. The rest of the cases would probably be covered by the daemon set approach indeed.

Reply:
Ironic also uses the MAC addresses to identify a host, so if it changes then it may be necessary to re-register the host or at least manually update the MAC in the host CRD and have the operator update ironic.